### ISSUE TITLE
Notification stuck at "Updating Weather information (0/1)" when using a current location

### ISSUE THREAD

**Issue by Penaz91** [breezy-weather/breezy-weather#193]
### Steps to reproduce

1. Set the automatic weather location to "Open-Meteo"
2. Set the weather provider to "Open-Meteo"
3. Pull down to update information

### Expected behavior

Weather information is either updated or an error is returned

### Actual behavior

Screen with only the background is shown (Title is "Automatic Position"), a notification saying "Updating Weather information" is in the notification drawer, stuck at "(0/1)"

### Weather provider used

Open-Meteo

### Breezy Weather version

4.2.1-beta

### Android version

Android 10 w/ MIUI 12.5.1 Global

### Device

Xiaomi RedMi Note 7

### Other details

On versions 4.1.2-beta and 4.1.3-beta everything was working correctly. The notification bar shows the usage of GPS location for a few seconds before stopping in all versions.

It also seems that the "Permanent notification" method of keeping the app alive is not available anymore (which is sad cause I can't manage to avoid MIUI from killing Breezy Weather and/or stopping from auto-starting).

I'm willing to help as much as possible in getting logs out of my mobile if necessary.

Thank you for your attention.

### Acknowledgements

- [X] I have searched the existing issues and this is a new ticket, **NOT** a duplicate or related to another open or closed issue.
- [X] I have written a short but informative title.
- [X] I have updated the app to **[latest version](https://github.com/breezy-weather/breezy-weather/releases/latest)**.

**Comment by papjul** [comment]
The new implementation is a mix of Android worker method with a notification to keep it alive during processing.
To avoid killing, please follow these steps:
https://dontkillmyapp.com/xiaomi

I seem to understand you are using current location:
1) What is your location provider? Android Native, or Baidu IP?
2) Are background updates working when you only add a city manually from the search?

**Comment by Penaz91** [comment]
1. My location provider is Android Native API
2. It seems that updates work well enough, as long as I remove the "Automatic Position" location from the list.

**Comment by Penaz91** [comment]
> The new implementation is a mix of Android worker method with a notification to keep it alive during processing. To avoid killing, please follow these steps: https://dontkillmyapp.com/xiaomi

For this it seems I had to uninstall Xiaomi's aggressive application killing (com.xiaomi.powerkeeper). Thanks!

**Comment by grrrrr** [comment]
Not sure if this warrants a separate ticket, I can create one if needed.    

I'll add I am seeing similar following the upgrade to v4.2.1-beta, prior versions worked as expected.

**Device**: Pixel 6a running GrapheneOS (AOSP 13) with no Google services installed.
**Location Provider:** NativeAPI
**Weather Service:** OpenWeather for bookmarked locations and current location.
**Location Access:** Always allowed (for v4.2.1 / for older versions this was "when used")
**Battery Usage:** Unrestricted (for v4.2.1 / for older versions this was "optimised")

Manually added cities are a bit hit and miss.   They don't seem to auto update and some times on a manual refresh they get stuck with a spinning icon.    Switching to another bookmarked location and a manual refresh gets an update

Background refresh does not appear to work.

**Comment by papjul** [comment]
> **Edit**: Sadly I have to take back what I said, even though Breezy Weather started automatically, it was somehow killed and the notification disappeared. I did everything on the website you mentioned (besides app pinning which for some reason doesn't seem to be available)

The notification is expected to disappear once it completes the job.


> Manually added cities are a bit hit and miss. They don't seem to auto update and some times on a manual refresh they get stuck with a spinning icon. Switching to another bookmarked location and a manual refresh gets an update

If you don't use a custom API key, OpenWeather is known to be rate-limited. However, it should show a toast message. This is indeed a different issue than the one with current location, so I need a different issue please ;)


> Background refresh does not appear to work.

Please check the presence of "WeatherUpdate-auto" in Settings > Background Updates > Worker info. If it's not there, please open a new issue.


> Switching the determination of the location on the manual one fixes the problem from time to time.

I don't understand what this means. Are you talking about current location or manually added cities? What does “from time to time” means?

**Comment by Penaz91** [comment]
> The notification is expected to disappear once it completes the job.

I realize I may have been misleading, my apologies. The notification I was talking about was the "permanent notification" that you can enable from "Settings -> Widget and animated background -> Notification Widget" (I hope I'm re-translating it correctly).

I think at this point the fact that the application doesn't even start automatically, even though all permissions are enabled, auto-start is enabled, etc... warrants a different issue, so I'll wait for @papjul 's confirmation on that and keep this issue focused on the weather update + geolocation issues.

**Comment by grrrrr** [comment]
> Please check the presence of "WeatherUpdate-auto" in Settings > Background Updates > Worker info. If it's not there, please open a new issue.

@papjul What I have is:

![breezy3](https://github.com/breezy-weather/breezy-weather/assets/1025506/1cdee7e0-6076-4960-ac7d-10c62052c5ed)

**Comment by papjul** [comment]
> > The notification is expected to disappear once it completes the job.
> 
> I realize I may have been misleading, my apologies. The notification I was talking about was the "permanent notification" that you can enable from "Settings -> Widget and animated background -> Notification Widget" (I hope I'm re-translating it correctly).

And this notification-widget worked flawlessly on v4.1.3?
I don’t remember changing anything to the code that would trigger MIUI killer “feature”.

> I think at this point the fact that the application doesn't even start automatically, even though all permissions are enabled, auto-start is enabled, etc... warrants a different issue, so I'll wait for @papjul 's confirmation on that and keep this issue focused on the weather update + geolocation issues.

What do you mean “application doesn't even start automatically”? It’s not supposed to.
If you mean the background job, can you see "WeatherUpdate-auto" in Settings > Background Updates > Worker info?

**Comment by JustMarkov** [comment]
> I don't understand what this means. Are you talking about current location or manually added cities? What does “from time to time” means?

Woops, sorry. I mean, that manually added cities seems to be showing the weather, but still not updating in the background. "From time to time” means that sometimes page refreshing updates weather, but sometimes not and requires to add city manually again

**Comment by papjul** [comment]
> > And this notification-widget worked flawlessly on v4.1.3?
> 
> This notification worked and I could "close" the app, leaving it running in the background (not using the Android Worker system though, but using the Permanent Notification Method).

OK, I will check if I did any change by mistake. However, it is not related at all to the Permanent Notification Method.


> Now it works as long as I keep the application open in my "recent apps" drawer.

This is a MIUI issue. Google forbids Android OEM to kill apps when they are removed from “recent apps”. Try again to find a way to set up the App pinning / App locking fix from here: https://dontkillmyapp.com/xiaomi


> > What do you mean “application doesn't even start automatically”? It’s not supposed to.
> > If you mean the background job, can you see "WeatherUpdate-auto" in Settings > Background Updates > Worker info?
> 
> The previous version was set to start automatically at the boot of my Android device (from the system settings), thus giving me the Weather Notification/Widget in my notification drawer without having to open the app explicitly. If the app has no such feature, then I may have confused unintended behaviour for a "feature", in that case I apologize.

Ah yes, you are right about that, I removed it because scheduled workers persist across reboots, but indeed notification-widget does not, as it is not a scheduled job. I will bring it back, sorry!


> I can see an Enqueued WeatherUpdate-Auto worker in the Settings -> Background Updates -> Worker Info.

That's a good thing, it means the job is now in the hands of Android to execute. The job is only created on first install/on updates. So if you see it there, it means that while MIUI kill the app, it does keep the scheduled job :)

You may be interested in following #194 that will allow you to see when background update was last executed so you can confirm it is indeed executed ;)

**Comment by papjul** [comment]
Do you use widgets?
If you do not, I think they should help keeping the app alive, just like a persistent notification would do. I know it's just a workaround but let me know if that helps!

**Comment by papjul** [comment]
> > I don't understand what this means. Are you talking about current location or manually added cities? What does “from time to time” means?
> 
> Woops, sorry. I mean, that manually added cities seems to be showing the weather, but still not updating in the background.

Sorry, missed your message, I really think you should open different issues :S

Just like with others, can you see "WeatherUpdate-auto" in Settings > Background Updates > Worker info? This will tell you if background updates job is scheduled (enqueued).

> "From time to time” means that sometimes page refreshing updates weather, but sometimes not and requires to add city manually again

But what happens exactly when you swipe to refresh? Can you make a screen recording?

**Comment by kosmoz** [comment]
To add another point of data, I also noticed that wheather updates don't seem to work for the current location anymore, except when using AccuWeather. Curiously, other providers still work for manually added cities, although communication with the worker seems to be a little wonky at times.

I don't think this is a battery optimization issue, as I disabled it for this app and generally don't have many problems with it on my phone (Fairphone 4, Android 12).

I'll try to reproduce this using a debug build this weekend if I have some time.

**Comment by Penaz91** [comment]
> Do you use widgets?
> If you do not, I think they should help keeping the app alive, just like a persistent notification would do. I know it's just a > workaround but let me know if that helps!

It seems that the widget is able to pull the last updated data from the application, but it does not update every 30 minutes (I set it to the shortest interval for testing purposes), only when I explicitly open the application.

> Actually, it should be added back once the first scheduled work is complete. So it won’t be there right away after reboot, > as we leave Android decide when it’s the best time to execute (it’s not a good thing that all apps execute their job at boot > time as it makes the device very busy), but if your refresh rate is 1 h 30 min, then it should be done no later than 1 h 40 after boot completed (we have a 10 min backoff policy).

Sadly this doesn't seem to be true, or at least it is mercurial at best, out of the last 6 or 7 reboots the notification appeared only once about 10 minutes after boot (using the 30 minute update rule).

> To add another point of data, I also noticed that wheather updates don't seem to work for the current location anymore, > except when using AccuWeather. Curiously, other providers still work for manually added cities, although communication with the worker seems to be a little wonky at times.

I almost forgot about this (which was the main point of the issue)! Thanks for reminding me! Here's a screen recording of the issue.

https://github.com/breezy-weather/breezy-weather/assets/6682630/574ba41d-ce62-4c2c-ba83-788d0a4a130b

**Comment by papjul** [comment]
> > Actually, it should be added back once the first scheduled work is complete. So it won’t be there right away after reboot, as we leave Android decide when it’s the best time to execute (it’s not a good thing that all apps execute their job at boot time as it makes the device very busy), but if your refresh rate is 1 h 30 min, then it should be done no later than 1 h 40 after boot completed (we have a 10 min backoff policy).
> 
> Sadly this doesn't seem to be true, or at least it is mercurial at best, out of the last 6 or 7 reboots the notification appeared only once about 10 minutes after boot (using the 30 minute update rule).

You rebooted 6-7 times with a 30 min interval every time? What does your notification history says exactly?

Regarding your video, it seems to me you're using a VPN, this could be a reason why it only happens to some people. If that's the case, I need to investigate why it gets stuck in that case.

**Comment by Penaz91** [comment]
> You rebooted 6-7 times with a 30 min interval every time? What does your notification history says exactly?

I rebooted 6-7 times with a 30 minute weather refresh interval set in the app, I restarted the phone about once every hour or more.

The notification log shows only notifications from breezy at 11.25am  and 10:26am (both times I manually opened the app), no other notifications since.

> Regarding your video, it seems to me you're using a VPN, this could be a reason why it only happens to some people. If that's the case, I need to investigate why it gets stuck in that case.

The VPN is just AdAway, even by disabling it the issue persists.

**Comment by papjul** [comment]
Thank you for taking time to do these tests and giving detailed replies, it helps a lot!

So, you have a widget that starts the app but doesn't trigger the scheduled job, but as soon as you open the app from its shortcut, scheduled job is triggered in **background** (with the notification progress), am I correct?

**Comment by jukkyrjo** [comment]
Pixel 6a. Only AccuWeather can be updated manually. None update in the background. Battery settings etc all fine

**Comment by papjul** [comment]
> Only AccuWeather can be updated manually. None update in the background.

Manually added cities or current location?

**Comment by Penaz91** [comment]
> So, you have a widget that starts the app but doesn't trigger the scheduled job, but as soon as you open the app from its shortcut, scheduled job is triggered in background (with the notification progress), am I correct?

Essentially yes, and while the app is open, the permanent notification I enabled is shown.

> Thank you for taking time to do these tests and giving detailed replies, it helps a lot!

No worries, I try to do what I can, I'm willing to use my device's USB debug mode to help, if needed :)

**Comment by jukkyrjo** [comment]
> > Only AccuWeather can be updated manually. None update in the background.
> 
> Manually added cities or current location?

Current location

**Comment by papjul** [comment]
f27e0e9d922f9713154d7707847a1623c7540bdb brings back BootReceiver

**Comment by seniorm0ment** [comment]
This seems to be related to location services. I cannot manually update even within the app after this past update, the current location.
If I manually search a city it works. But current location just fails.

**Comment by papjul** [comment]
> This seems to be related to location services. I cannot manually update even within the app after this past update, the current location. If I manually search a city it works. But current location just fails.

Which provider do you use?

**Comment by seniorm0ment** [comment]
Accuweather

**Comment by papjul** [comment]
We discussed a lot of matters in this issue. The original issue regarding weather update being stuck is now fixed in v4.3.0-beta:
https://github.com/breezy-weather/breezy-weather/releases/tag/v4.3.0-beta

If you experience other issues than that, please open a new issue.
Following your helpful feedback, if there are no major regressions, I will move v4.3.0 from pre-release to latest release.
